uaclient/files.py

```diff
@@ -0,0 +1,296 @@
+from typing import Dict, Optional
+
+from uaclient import defaults, event_logger, schemas, util
+
+PRIVATE_SUBDIR = "private"
+
+event = event_logger.get_event_logger()
+LOG = logging.getLogger(__name__)
+
+
+class UAFile:
+    _root_directory = defaults.DEFAULT_DATA_DIR
+    _file_name = "file"
+
+    def __init__(
+        self,
+        root_directory: Optional[str] = None,
+        filename: Optional[str] = None,
+        private: bool = True,
+    ):
+        if root_directory:
+            self._root_directory = root_directory
+        if filename:
+            self._file_name = filename
+        self._is_private = private
+        self._filepath = None
+
+    @property
+    def filepath(self):
+        if not self._filepath:
+            self._filepath = os.path.join(
+                self._root_directory, self._file_name
+            )
+        return self._filepath
+
+    @property
+    def is_private_file(self):
+        return self._is_private
+
+    def write(self, content):
+        file_mode = 0o600 if self._is_private else 0o644
+        self.create_file()
+        if not isinstance(content, str):
+            content = json.dumps(content, cls=util.DatetimeAwareJSONEncoder)
+        util.write_file(self.filepath, content, file_mode)
+
+    def read(self):
+        content = None
+        try:
+            content = util.load_file(self.filepath)
+        except FileNotFoundError:
+            if not os.path.exists(self.filepath):
+                LOG.debug("File does not exist: {}".format(self.filepath))
+        return content
+
+    def delete(self):
+        util.remove_file(self.filepath)
+
+    def create_file(self):
+        file_dir = os.path.dirname(self.filepath)
+        if not os.path.exists(file_dir):
+            os.makedirs(file_dir)
+            if self._is_private:
+                os.chmod(file_dir, 0o700)
+
+
+class PublicMachineTokenFile(UAFile):
+    _file_name = defaults.PUBLIC_MACHINE_TOKEN_FILE
+    _is_private = False
+
+    def __init__(
+        self,
+        root_directory: Optional[str] = None,
+    ):
+        if root_directory:
+            self._root_directory = root_directory
+        file_name = defaults.PUBLIC_MACHINE_TOKEN_FILE
+        is_private = False
+        super().__init__(self._root_directory, file_name, is_private)
+
+    def write(self, content):
+        """
+        Writes to the public file of the machine token
+        It checks for new keys and filter the data
+        """
+        try:
+            if isinstance(content, str):
+                content = json.loads(content)
+        except Exception:
+            return None
+        self.filter_tokens(schemas.contract_schema, content)
+        content = json.dumps(content, cls=util.DatetimeAwareJSONEncoder)
+        content = content.replace('"', "'")
+        content = util.redact_sensitive_logs(content)
+        content = content.replace("'", '"')
+        super().write(content)
+
+    def filter_tokens(self, schema, data):
```

Comment:
There might be a way to use DataObjects for this instead of the schema/example dict.

I think you could define the schema with DataObjects, and then just do a `PublicMachineTokenData.from_dict(d).to_dict()` and that would filter out the disallowed fields.